### ISSUE TITLE
[Feature Request] efficient logdet for KroneckerProductLazyTensor

### ISSUE THREAD

**Issue by mshvartsman** [cornellius-gp/gpytorch#725]
# 🚀 Feature Request

It appears that KroneckerProductLazyTensor doesn't compute efficient logdets (as a product of the scaled log-determinants of the factors). This means it costs O(n^2m^2) for a 2-factor kronecker kernel with n x n and m x m factors instead of O(n^2 + m^2) even if we're using efficient inv_quads. 

This is fairly straightforward to add and would speed up multitask models and probably a bunch of other stuff. I'm happy to do it, just a bit confused on (a) what `reduce_inv_quad` does and (b) where the efficient inv_quad is happening right now and whether I can reuse it or also need to implement an efficient kron-solve. 

Simple example: 
```
import gpytorch
import torch
from gpytorch.functions import logdet, inv_matmul
from gpytorch.utils.cholesky import psd_safe_cholesky
LOG2PI = 1.8378770664093453

def lazy_matrixnormal_likelihood(X, R, C):
    """
    Likelihood for matrix-variate normal
    (https://en.wikipedia.org/wiki/Matrix_normal_distribution)
    using lazy logdet and inv_matmul
    """
    row_logdet = logdet(R)
    col_logdet = logdet(C)

    logdet_term = X.shape[0] * col_logdet + X.shape[1] * row_logdet

    trace_term = torch.trace(inv_matmul(R, X).matmul(inv_matmul(C, X.t())))

    return -0.5 * (X.shape[1] * X.shape[0] * LOG2PI + logdet_term + trace_term)


def lazy_kron_mvn(X, K):
    
    """likelihood for multivariate normal using lazy
    """
    logdet_term = logdet(K)

    trace_term = X.t().matmul(inv_matmul(K, X))
    return -0.5 * (X.shape[0] * LOG2PI + logdet_term + trace_term)


def chol_logdet(L):
    """
    log determinant of a matrix with cholesky factor L
    """
    return 2 * torch.sum(torch.log(torch.diag(L)))


def cholesky_solve(x, L):
    """
    Solve positive definite system using the cholesky factor 
    """
    y = torch.triangular_solve(x, L, upper=False).solution
    return torch.triangular_solve(y, L.t(), upper=True).solution

def nonlazy_matrixnormal_likelihood(X, R, C):
    cholr = psd_safe_cholesky(R)
    cholc = psd_safe_cholesky(C)

    logdet_term = X.shape[0] * chol_logdet(cholc) + X.shape[1] * chol_logdet(cholr)

    trace_term = torch.trace(cholesky_solve(X, cholr).matmul(cholesky_solve(X.t(), cholc)))

    return -0.5 * (X.shape[1] * X.shape[0] * LOG2PI + logdet_term + trace_term)
def kron_bench(n=100, m=100, device='cpu'):
    print(f"====== n={n},m={m}, device={device} ======")
    X = torch.arange(n, dtype=torch.float, device=device)[:,None] @ torch.arange(m, dtype=torch.float, device=device)[None,:]
    y = torch.randn(m, n, device=device)
    yflat = y.flatten()
    muflat = torch.zeros(n*m, device=device)
    k = gpytorch.kernels.RBFKernel()
    K_row = (k(X) + torch.eye(m)).to(device)
    K_col = (k(X.t()) + torch.eye(n)).to(device)
    K_kron = gpytorch.lazy.KroneckerProductLazyTensor(K_col, K_row).to(device)
    
    print(f"\n matrix-normal with lazy solve/logdet")
    print(f"logp: {lazy_matrixnormal_likelihood(y, K_row, K_col)}")
    %timeit lazy_matrixnormal_likelihood(y, K_row, K_col)

    print(f"\n matrix-normal with torch cholesky")
    print(f"logp: {nonlazy_matrixnormal_likelihood(y, K_row.evaluate(), K_col.evaluate())}")
    %timeit nonlazy_matrixnormal_likelihood(y, K_row.evaluate(), K_col.evaluate())
    
    print(f"\n multivariate normal with KronLazy")
    print(f"logp: {lazy_kron_mvn(yflat, K_kron)}")
    %timeit lazy_kron_mvn(yflat, K_kron)
    
    print(f"\n gpytorch multivariate normal with Kronlazy")
    mvn_distr = gpytorch.distributions.multivariate_normal.MultivariateNormal(mean=muflat, covariance_matrix=K_kron)
    print(f"logp: { mvn_distr.log_prob(yflat)}")
    %timeit mvn_distr.log_prob(yflat)    
    
    
    print("\n")
    

kron_bench(n=100, m=100, device='cpu')
kron_bench(n=100, m=100, device='cuda')
kron_bench(n=1000, m=1000, device='cpu')
kron_bench(n=1000, m=1000, device='cuda')
```

```
====== n=100,m=100, device=cpu ======

 matrix-normal with lazy solve/logdet
logp: -17354.173828125
1.55 ms ± 16.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

 matrix-normal with torch cholesky
logp: -17354.173828125
716 µs ± 11 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

 multivariate normal with KronLazy
logp: -17354.173828125
100 ms ± 2.16 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

 gpytorch multivariate normal with Kronlazy
logp: -17354.171875
98.9 ms ± 1.89 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


====== n=100,m=100, device=cuda ======

 matrix-normal with lazy solve/logdet
logp: -17373.841796875
8.63 ms ± 14.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

 matrix-normal with torch cholesky
logp: -17373.841796875
5.62 ms ± 14.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

 multivariate normal with KronLazy
logp: -17373.841796875
43.5 ms ± 23.9 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

 gpytorch multivariate normal with Kronlazy
logp: -17373.83984375
30.2 ms ± 27.6 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


====== n=1000,m=1000, device=cpu ======

 matrix-normal with lazy solve/logdet
logp: -1737016.125
820 ms ± 7.41 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

 matrix-normal with torch cholesky
logp: -1737016.25
40.1 ms ± 420 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

 multivariate normal with KronLazy
logp: -1737016.0
16.2 s ± 157 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

 gpytorch multivariate normal with Kronlazy
logp: -1737015.75
17.3 s ± 118 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


====== n=1000,m=1000, device=cuda ======

 matrix-normal with lazy solve/logdet
logp: -1735884.75
87 ms ± 199 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

 matrix-normal with torch cholesky
logp: -1736324.125
23.9 ms ± 892 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

 multivariate normal with KronLazy
logp: -1736284.5
570 ms ± 194 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

 gpytorch multivariate normal with Kronlazy
logp: -1736323.625
578 ms ± 118 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by jacobrgardner** [comment]
The correct performance comparison would be with `LazyTensor.inv_quad_log_det`, which computes both terms simultaneously, hiding the logdet computation behind the solve. It’s worth noting that the running time of `inv_quad`, `logdet` and `inv_quad_log_det` should all be `O(n^2m + m^2n)` — this seems like it might end up being a perf bug.

If it turns out that it is faster to compute the log det separate from the `inv_quad` term, this could be accomplished by overriding `inv_quad_log_det` to compute the `inv_quad` term only with the `InvQuadLogDet` function, and then compute log dets; but let’s ensure that this isn’t a perf bug first.

**Comment by jacobrgardner** [comment]
With that said, I agree at a minimum that `KroneckerProductLazyTensor.logdet` should probably be overridden so that if users aren’t asking for a linear solve / inv quad as well, we do the fast thing. It’ll be interesting to see whether it should also be used in the `inv_quad_logdet` case.

If it is indeed faster in the `inv_quad_logdet` case, then what I would do is just never pass the `logdet=True` on to the `inv_quad_logdet` function, and have an `if logdet:` branch that just calls `self.logdet()`, where `KroneckerProductLazyTensor.logdet` has been written to do the right thing.

**Comment by mshvartsman** [comment]
Point well-taken re: the correct complexity (not used to worrying about the matmul costs given the cost of the cholesky, oops). 

Re: the correct performance comparison, I think the single call to `inv_quad_logdet` is what `gpytorch.distributions.multivariate_normal.MultivariateNormal` does, which is my final comparison above (and has similar performance to my two-call variant). Based on this I think the right solution is to implement `KroneckerProductLazyTensor.inv_quad_logdet` that collects the logdets as it does the matmuls for each factor, and applies the identity to give us a fast logdet in both calls. Thoughts?

**Comment by Balandat** [comment]
I think that is exactly what we'd want to do. Ideally it's possible to compute the individual logdets in batch mode, but sequential should still be a big improvement.

**Comment by mshvartsman** [comment]
Sounds good, I'll try to poke at it this week unless someone else thinks they may get to it?

**Comment by jacobrgardner** [comment]
If you want to take a stab at it, that would be great! Otherwise I can try to take a look at it sometime in the next two weeks.

**Comment by mshvartsman** [comment]
Super gnarly code that might break autodiff, but just verifying we're talking about the same thing here: 

```
import torch
import gpytorch
import numpy as np

class KronWithFastLogDet(gpytorch.lazy.KroneckerProductLazyTensor):
    
    def logdet(self):
        logdets = torch.Tensor([lt.logdet() for lt in self.lazy_tensors]) 
        tensor_shapes = [x.shape[0] for x in self.lazy_tensors]
        multipliers = torch.Tensor([torch.prod(torch.Tensor([ts for i, ts in enumerate(tensor_shapes) if i!=j]))
                                    for j in range(len(tensor_shapes))])
        return torch.sum(logdets * multipliers)
    
    
def kronlogdet_bench(*sizes):

    k = gpytorch.kernels.RBFKernel()
    mats = [k(torch.arange(n, dtype=torch.float)) + torch.eye(n) for n in sizes]
    K_kron = gpytorch.lazy.KroneckerProductLazyTensor(*mats)
    K_kronfast = KronWithFastLogDet(*mats)
    print("\n default ")
    %timeit K_kron.logdet()
    print("\n fast logdet")
    %timeit K_kronfast.logdet()
    print(f"values: default: {K_kron.logdet()}, fast:{K_kronfast.logdet()}, numpy:{np.linalg.slogdet(K_kron.numpy())}")
    

# very small: we get identical answer but (I'm guessing) batching beats the trick?
kronlogdet_bench(2,3) 
# default 
# 274 µs ± 80.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

#  fast logdet
# 623 µs ± 92.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
# values: default: 8.094792366027832, fast:8.094792366027832, numpy:(1.0, 8.094792)

# bigger: way faster and the naive algo seems to disagree with fast and numpy; is this due to the approximation in the iterative solver?
kronlogdet_bench(20, 30) 
#  default 
# 28.2 ms ± 2.41 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

#  fast logdet
# 676 µs ± 104 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
# values: default: 796.61083984375, fast:794.685302734375, numpy:(1.0, 794.6852)

# really emphasize the difference with more factors
kronlogdet_bench(10, 10, 10, 10) 

#  default 
# 75.4 ms ± 13.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

#  fast logdet
# 1.4 ms ± 399 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# values: default: 26520.134765625, fast:26566.232421875, numpy:(1.0, 26566.23)
```

Are the numerical differences of the slow version relative to fast (or numpy) consistent with what we'd expect from the iterative algorithm being used? 

Where would I get into capturing the logdets as part of the solves of the individual factors? I think that's hidden behind the `self._matmul` being passed into `self._solve`, and I can't figure out how to get into that without touching `InvQuadLogDet` itself. I'm not sure how much extra work the clever version is once someone tells me where things go :) -- I could just yank the list comprehensions and PR a cleaner version of the above pretty quickly (though I might need a hand verifying that I'm handling all the batch semantics correctly).

**Comment by jacobrgardner** [comment]
I think those deviations look reasonable assuming a default number of trace samples (e.g., around a 0.2% error).

Getting this change in is slightly tricky because we want `InvQuadLogDet` to still do the right thing in the backwards pass. Basically, there are two if statements in the forward of `InvQuadLogDet` that check if `ctx.logdet` is true:
https://github.com/cornellius-gp/gpytorch/blob/4c71e54f519fe1c51db8d0609c5a88532039f17b/gpytorch/functions/_inv_quad_log_det.py#L95-L97
and
https://github.com/cornellius-gp/gpytorch/blob/4c71e54f519fe1c51db8d0609c5a88532039f17b/gpytorch/functions/_inv_quad_log_det.py#L122-L134

What I think we should do is have `KroneckerProductLazyTensor` define your `logdet` method as written, and then a `_logdet_override` method that calls and returns it. Then, in `InvQuadLogDet.forward`, if `lazy_tsr` has a `_logdet_override` method defined, we call it and return set the result to `logdet_term`, and the behavior of neither if statement happens (e.g., we neither augment the rhs_list in the first if statement nor do any of the eigenvalue decompositions in the second if statement).

**Comment by jacobrgardner** [comment]
Acutally, correction: we do need the first `if ctx.logdet` if statement to happen (e.g., the augmentation of the `rhs_list`). This is because we still need solves w.r.t the probe vectors in the backward pass.

This may end up negating some of the time savings from this when using to train a GP, because we still have to do additional linear solves to get the backward.

**Comment by mshvartsman** [comment]
If I understand this correctly, then your suggestion would still do the solves once for the inv_quad (using the fast matmul and CG) and another time for the logdets (using separate solves for each factor). It seems like I could do this in a simpler way than getting into InvQuadLogDet at all: 

```
class KronWithFastLogDet(gpytorch.lazy.KroneckerProductLazyTensor):
    
    def inv_quad_logdet(self, inv_quad_rhs=None, logdet=False, reduce_inv_quad=True):
        inv_quad_term = super().inv_quad_logdet(inv_quad_rhs=inv_quad_rhs, logdet=False, reduce_inv_quad=reduce_inv_quad)
        if logdet: 
            logdet_term = self.logdet()
        else: 
            logdet_term = 0
        return inv_quad_term, logdet_term
    
    def logdet(self):
        logdets = torch.Tensor([lt.logdet() for lt in self.lazy_tensors]) 
        tensor_shapes = [x.shape[0] for x in self.lazy_tensors]
        multipliers = torch.Tensor([torch.prod(torch.Tensor([ts for i, ts in enumerate(tensor_shapes) if i!=j]))
                                    for j in range(len(tensor_shapes))])
        return torch.sum(logdets * multipliers)

```

Does this make sense? It also seems like theoretically we should be able to reuse those solves, by wrapping the same logic that does matmul around them instead of passing the fast matmul into the CG solver: 

```
class KronWithFastInvQuadLogdet(gpytorch.lazy.KroneckerProductLazyTensor):
    
    def inv_quad_logdet(self, inv_quad_rhs=None, logdet=False, reduce_inv_quad=True):

        nrows, ncols = inv_quad_rhs.shape

        nfactors = len(self.lazy_tensors)

        logdets = torch.zeros(nfactors)
        tensor_shapes = [q.shape[0] for q in self.lazy_tensors]
        multipliers = torch.Tensor([torch.prod(torch.Tensor([ts for i, ts in enumerate(tensor_shapes) if i!=j]))
                                for j in range(len(tensor_shapes))])

        y = inv_quad_rhs.clone()
        for i, q in enumerate(self.lazy_tensors):
            n = q.shape[0]
            # ideally we could just get inv_matmul and logdet from the one call instead of
            # two here, reusing solves the same way we do in usual InvQuadLogDet
            y = q.inv_matmul(y.contiguous().view(n, -1))
            logdets[i] = q.logdet()
            y = y.view(n, nrows//n, -1).permute(1, 0, 2)

        logdet_term = torch.sum(logdets * multipliers)

        inv_quad_term = torch.trace(inv_quad_rhs.t() @ y.contiguous().view(nrows, ncols))

        return inv_quad_term, logdet_term
```

In my testing (major caveat: all on CPU right now):
- for small kronecker factors (e.g. [20, 30]), you get something like a 10x speedup with the second version above relative to either the naive version or the first version above. This is in the domain where we end up doing choleskys on the individual factors vs CG using the efficient kron matmul.
- for larger kronecker factors (e.g. [130, 130]), you get a more modest speedup (maybe 1.5-2x). This is in the domain where we're always doing CG but either using the fast kronmult or using the fast kronsolve. 

So I guess: 
- Should I contribute one of the versions above, or can you help me understand why doing this in InvQuadLogDet as you describe is preferred? 
- Would you consider optionally returning `inv_quad_solves` out of `InvQuadLogDet` so we can do an even more efficient version?

**Comment by jacobrgardner** [comment]
Sorry for the delay reviewing this, I've been really sick and also at ICML for the past few days!

I agree that modifying the method on KroneckerProductLazyTensor is the way to go given that the first if statement adding the right hand sides can't be bypassed. Therefore, the second approach makes the most sense to me.

Regarding modifications to `InvQuadLogDet`: I'm pretty much fine with anything that results in substantial efficiency gains; however, note that returning additional things out of `Functions` can sometimes be a headache because PyTorch might expect you to supply gradients for them in `backward`.

Once I get back in front of a real computer, I'd be happy to benchmark this on the GPU as well just to confirm we don't see any speed losses there as well.

Thanks for being willing to take this on!

**Comment by mshvartsman** [comment]
No worries re delay, I'm just monkeypatching my hackaround locally while we're figuring this out. I think at this stage let's get the change without modifying `InvQuadLogDet` done (this would address the majority of the performance gap), and worry about those modifications later. FWIW in my naive testing (where I basically called the CG solver directly and tried replicating the logic in `InvQuadLogDet` by hand to reuse the solves) it went slower rather than faster, not sure what I was doing wrong. 

To get this done properly, I think my version above won't work since it doesn't handle batched inputs (also to @Balandat's point). But I'm not sure I understand all the batch semantics. It seems like: 
- KroneckerProductLazyTensor could be batched, which means each element in `self.lazy_tensors` is `[N, N]` or `[B, N, N]`.  
- `inv_quad_rhs` could be batched, which means it is `[N, M]` or `[B, N, M]`. 
- `inv_quad_term` should return something that is either a scalar or tensor of dimension `B` and similarly for `logdet_term`. 

Therefore, I'm imagining the right thing is something like figuring out if we're batched or not, adding a dimension B=1 if needed, doing all the logic as if we're batched (i.e. always using negative indices on reshapes/views), and then removing the fake batch if we weren't batched. Is that about right? 

I think there's a bunch of machinery in LazyTensor that helps with the above (adding the batch dimension, correctly indexing into the batch vs nonbatch dimensions etc) and I've started digging to understand where it lives, but I'd appreciate signposts on where/how to not reinvent the wheel. I'm a little lost in the abstraction :).

**Comment by jacobrgardner** [comment]
@mshvartsman do you have a sense of whether `KronWithFastLogDet` gets most of the speed of `KronWithFastInvQuadLogDet`, or do we need to modify the latter? I'm happy to help get this in now. 

Batch mode support is usually straightforward to modify for, because ops like matmuls etc maintain the batch shape of their input tensors. Usually supporting batch mode boils down to two key principles: (1) use relative dimension indices (e.g. -2 instead of 0), and (2) when doing view/expand/etc calls, just prepend a `*batch_shape` to the front of the call, where `batch_shape` might be `some_relevant_tensor.shape[:-2]`

**Comment by mshvartsman** [comment]
I think for my case (a pile of mostly small factors) going from a big CG to a bunch of small choleskys is a huge boost because of the relatively big constant cost in CG, but not sure how this looks when you have bigger factors (without thinking deeply, rough intuition is we'll top out at ~2x since the efficient matmul looks a lot like the efficient kronsolve, so the savings is mainly in reusing the solve+logdet call; maybe a bigger boost if we can figure out how to batch things cleverly). Can test but am slammed and probably won't get to think more about this in detail until late next week, sorry. Pointer to relative dim indices and batch shape is perfect though, thank you. 

Also: wanted to flag for anyone coming across this thread and using my little snippets above that there's a small bug. You need to make sure to send the `logdets` and `multiplier` tensors to the correct device, else forward works correctly but the backward call fails. `multipliers = multipliers.to(self.lazy_tensors[0].device)` and same for `logdets` is a simple way to do it.

**Comment by Balandat** [comment]
Looks like this is still open and will require `inv_quad_logdet` `KroneckerProductLazyTensor`? Currently the efficient Cholesky will be used if the matrix is small enough, but not for larger matrices (where this would be useful).

**Comment by wjmaddox** [comment]
This should be finally resolved by #1786.
